Generated attribute setters for nullable types fail when passing null.

rreno opened the pull request:
#### 23551a03b824a73d7f0b63db00a35706701324f5
<pre>
Generated attribute setters for nullable types fail when passing null.
<a href="https://bugs.webkit.org/show_bug.cgi?id=259947">https://bugs.webkit.org/show_bug.cgi?id=259947</a>
rdar://113591277

Reviewed by Chris Dumez.

We aren&apos;t allowing `null` as an attribute value in the setter for
nullable enum attributes. This change will check if the attribute is
nullable in the code generator and if so generate C++ code which will
call the native implementation of the setter with `std::nullopt` when
passed `null` in the JS code.

Preserves the existing behavior for non-nullable attributes.

* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateAttributeSetterBodyDefinition):
* Source/WebCore/bindings/scripts/test:
    Added an attribute to TestObj to check the output of the generator for
    nullable enum type attributes.

Canonical link: <a href="https://commits.webkit.org/266726@main">https://commits.webkit.org/266726@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36b75155e96ea858ced4b70107ebab6369017088

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14568 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14879 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15223 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16315 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13764 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14704 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17393 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14956 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16423 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14750 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15268 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12371 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17045 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12551 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13140 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20141 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13630 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13305 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16539 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13856 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-WK2-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13149 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3522 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17484 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13700 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5683,13 +5683,15 @@ sub GenerateAttributeSetterBodyDefinition
-            push(@$outputArray, "    if (UNLIKELY(!optionalNativeValue))\n");
-            push(@$outputArray, "        return false;\n");
-            push(@$outputArray, "    auto nativeValue = optionalNativeValue.value();\n");
+            if (!$attribute->type->isNullable) {
```
This is not correct. We should not be calling parseEnumeration() at all if the input is null or undefined.

Now if the input is an invalid string, you'll implicitly convert to null which is not spec compliant.

cdumez commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5681,10 +5681,26 @@ sub GenerateAttributeSetterBodyDefinition
     } else {
         push(@$outputArray, "    auto& impl = thisObject.wrapped();\n") if !$attribute->isStatic;
-       
+
```
I think it would be nice to add an anonymous function here to avoid code duplication:
```
my $generateFunctionString = sub {
    my $nativeValue = shift;

    my ($baseFunctionName, @arguments) = $codeGenerator->SetterExpression(\%implIncludes, $interface->type->name, $attribute);
    push(@arguments, $nativeValue);
    my $functionName = GetFullyQualifiedImplementationCallName($interface, $attribute, $baseFunctionName, "impl", $conditional);
    AddAdditionalArgumentsForImplementationCall(\@arguments, $interface, $attribute, "impl", "lexicalGlobalObject", "", "thisObject");
    return "${functionName}(" . join(", ", @arguments) . ")";
};
```

rreno replied:
I didn't know about anonymous functions in Perl - pretty cool. Though the syntax for actually calling it was a bit unexpected (`$generateFunctionString->("std::nullopt");`)

cdumez replied:
I think `&$generateFunctionString("std::nullopt");` would work too but yes, it's Perl :/

cdumez commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5681,10 +5681,26 @@ sub GenerateAttributeSetterBodyDefinition
+                my $functionName = GetFullyQualifiedImplementationCallName($interface, $attribute, $baseFunctionName, "impl", $conditional);
+                AddAdditionalArgumentsForImplementationCall(\@arguments, $interface, $attribute, "impl", "lexicalGlobalObject", "", "thisObject");
+
+                my $functionString = "${functionName}(" . join(", ", @arguments) . ")";
```
`my $functionString = generateFunctionString("std::nullopt");`

cdumez commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5681,10 +5681,26 @@ sub GenerateAttributeSetterBodyDefinition
+                push(@$outputArray, "        return true;\n");
+                push(@$outputArray, "    }\n\n");
+            }
+
```
And this at line 5718:
```
`my $functionString = generateFunctionString(PassArgumentExpression("nativeValue", $attribute));`
```

cdumez reviewed (approved):
LGTM

rreno committed:
Generated attribute setters for nullable types fail when passing null.
https://bugs.webkit.org/show_bug.cgi?id=259947
rdar://113591277

Reviewed by Chris Dumez.

We aren't allowing `null` as an attribute value in the setter for
nullable enum attributes. This change will check if the attribute is
nullable in the code generator and if so generate C++ code which will
call the native implementation of the setter with `std::nullopt` when
passed `null` in the JS code.

Preserves the existing behavior for non-nullable attributes.

* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateAttributeSetterBodyDefinition):
* Source/WebCore/bindings/scripts/test:
    Added an attribute to TestObj to check the output of the generator for
    nullable enum type attributes.

Canonical link: https://commits.webkit.org/266726@main
